addons/survey/models/survey.py

```diff
@@ -0,0 +1,1066 @@
+
+    @api.multi
+    def _read_group_stage_ids(self, domain, read_group_order=None, access_rights_uid=None):
+        """ Read group customization in order to display all the stages in the
+        kanban view, even if they are empty """
+        SurveyStage = self.env['survey.stage']
+        order = SurveyStage._order
+        access_rights_uid = access_rights_uid or self.env.uid
+
+        if read_group_order == 'stage_id desc':
+            order = '%s desc' % order
+
+        stage_ids = SurveyStage._search([], order=order, access_rights_uid=access_rights_uid)
+        stages = SurveyStage.sudo(access_rights_uid).browse(stage_ids)
+        result = stages.name_get()
+
+        # restore order of the search
+        result.sort(lambda x, y: cmp(stage_ids.index(x[0]), stage_ids.index(y[0])))
+
+        fold = {}
+        for stage in stages:
+            fold[stage.id] = stage.fold
+        return result, fold
+
+    _group_by_full = {
+        'stage_id': _read_group_stage_ids
+    }
+
+    # Actions
+
+    @api.multi
+    def action_start_survey(self):
+        ''' Open the website page with the survey form '''
+        self.ensure_one()
+        trail = ""
+        context = dict(self.env.context, relative_url=True)
+        if 'survey_token' in context:
+            trail = "/" + context['survey_token']
+        return {
+            'type': 'ir.actions.act_url',
+            'name': "Start Survey",
+            'target': 'self',
+            'url': self.with_context(context).public_url + trail
+        }
+
+    @api.multi
+    def action_send_survey(self):
+        ''' Open a window to compose an email, pre-filled with the survey
+        message '''
+        self.ensure_one()
+        if not self.designed:
+            raise UserError(_('You cannot send an invitation for a survey that has no questions.'))
+
+        if self.stage_id.closed:
+            raise UserError(_("You cannot send invitations for closed surveys."))
+        template_id = self.env.ref('survey.email_template_survey', raise_if_not_found=False)
+        ctx = dict(self.env.context)
+        ctx.update({'default_model': 'survey.survey',
+                    'default_res_id': self.id,
+                    'default_survey_id': self.id,
+                    'default_use_template': bool(template_id),
+                    'default_template_id': template_id and template_id.id or False,
+                    'default_composition_mode': 'comment'}
+                   )
+        return {
+            'type': 'ir.actions.act_window',
+            'view_type': 'form',
+            'view_mode': 'form',
+            'res_model': 'survey.mail.compose.message',
+            'target': 'new',
+            'context': ctx,
+        }
+
+    @api.multi
+    def action_print_survey(self):
+        ''' Open the website page with the survey printable view '''
+        self.ensure_one()
+        trail = ""
+        context = dict(self.env.context or {}, relative_url=True)
```

Comment:
remove 'or {}', even you can omit this context variable and use 'survey_token' in self.env.context  and at line 234 self.with_context(relative_url=True).print_url